drivers/wifi/tcp_socket.c

```diff
@@ -0,0 +1,181 @@
+ * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, NON-INFRINGEMENT,
+ * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+ * IN NO EVENT SHALL ANALOG DEVICES BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, INTELLECTUAL PROPERTY RIGHTS, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*******************************************************************************/
+
+/******************************************************************************/
+/***************************** Include Files **********************************/
+/******************************************************************************/
+
+#include <stdlib.h>
+#include "error.h"
+#include "tcp_socket.h"
+
+/******************************************************************************/
+/*************************** Types Declarations *******************************/
+/******************************************************************************/
+
+/* Socket descriptor */
+struct tcp_socket_desc {
+	/* Id of the opened socket */
+	uint32_t			id;
+	/* Reference to the network interface */
+	struct network_interface	*net;
+};
+
+/******************************************************************************/
+/************************ Functions Definitions *******************************/
+/******************************************************************************/
+
+/**
+ * @brief Allocate resources and initializes the socket descriptor
+ * @param desc - Address where to store the socket descriptor
+ * @param param - Initializing data
+ * @return
+ *  - \ref SUCCESS : On success
+ *  - \ref FAILURE : Otherwise
+ */
+int32_t socket_init(struct tcp_socket_desc **desc,
+		    struct tcp_socket_init_param *param)
+{
+	struct tcp_socket_desc	*ldesc;
+	int32_t			ret;
+
+	if (!desc || !param)
+		return FAILURE;
+
+	ldesc = (typeof(ldesc))calloc(1, sizeof(*ldesc));
+	if (!ldesc)
+		return FAILURE;
+
+	ldesc->net = param->net;
+
+	ret = ldesc->net->socket_open(ldesc->net->net, &ldesc->id,
+				      PROTOCOL_TCP);
+	if (IS_ERR_VALUE(ret)) {
+		free(ldesc);
+		return FAILURE;
+	}
+	*desc = ldesc;
+
+	return SUCCESS;
+}
+
+/**
+ * @brief Deallocate resources from the socket descriptor
+ * @param desc - Socket descriptor
+ * @return
+ *  - \ref SUCCESS : On success
+ *  - \ref FAILURE : Otherwise
+ */
+int32_t socket_remove(struct tcp_socket_desc *desc)
+{
+	int32_t ret;
+
+	if (!desc)
+		return FAILURE;
+
+	ret = desc->net->socket_close(desc->net->net, desc->id);
+	if (IS_ERR_VALUE(ret))
+		return FAILURE;
```

Comment:
We don't have good status/error code support in no-OS (yet), but I think for the future it would be better in these situations to forward `ret` back to caller. Sometimes it is indeed needed to return a specific error code, but many times, including here, it's ok to just pass `ret`. And When we will have more than SUCCESS and FAILURE you won't need to update the code here.

Reply:
returned ret